File: src/cli/tui/info.ts
/**
 * TUI runtime facts (GRACE full-screen interface).
 *
 * Every value in the header/home screen comes from real state: the workspace
 * the CLI was launched in, the configured provider/model, the auth session and
 * (best-effort) the server's free-plan quota. Nothing is invented.
 */
import { homedir } from 'node:os';
import { ApiClient } from '../../auth/client.ts';
import { loadSession, sessionExpired } from '../../auth/session.ts';
import { VERSION } from '../../meta.ts';
import { RemoteProvider } from '../../providers/remote.ts';
import type { Runtime } from '../../runtime.ts';
import { shortPath } from '../../util/text.ts';
import { stripAnsi } from '../ui/theme.ts';
import { bannerFreePlanLine, formatCountdown, sessionSecondsLeft } from '../freePlan.ts';
import type { TuiInfo } from './types.ts';

/**
 * Synchronous snapshot of real runtime facts. `freePlanLine` may be passed in
 * from an earlier async fetch (see refreshFreePlan).
 */
export function buildTuiInfo(runtime: Runtime, freePlanLine?: string): TuiInfo {
  const stored = loadSession();
  const loggedIn = stored !== null && !sessionExpired(stored);

  let sessionLabel: string;
  if (loggedIn && stored) {
    sessionLabel = stored.user.email;
  } else {
    sessionLabel = 'Local mode';
  }

  const served = runtime.provider instanceof RemoteProvider ? (runtime.provider.serverProvider ?? RemoteProvider.sharedServerProvider()) : null;
  const providerLabel = runtime.provider ? (served?.label ?? runtime.provider.label) : '';
  const model = runtime.provider ? runtime.provider.getModel().id : '';

  return {
    version: VERSION,
    workspace: shortPath(runtime.root, homedir()),
    provider: providerLabel,
    providerAvailable: runtime.provider !== null,
    providerError: runtime.providerError ?? undefined,
    model,
    session: sessionLabel,
    freePlan: freePlanLine,
  };
}

/**
 * Best-effort free-plan + session-time fetch from the backend (real data).
 * Never delays or breaks the UI: failures just leave the line unset.
 */
export async function refreshFreePlan(runtime: Runtime): Promise<string | undefined> {
  const stored = loadSession();
  if (!stored || sessionExpired(stored)) return undefined;
  if (!(runtime.provider instanceof RemoteProvider)) return undefined;
  try {
    const state = await new ApiClient(stored.apiUrl, 2000).getUsage(stored.token);
    // Seed the shared session view so the live countdown renders even before
    // the first task (display only — the server enforces the session limit).
    RemoteProvider.setSharedSession(state);
    return stripAnsi(bannerFreePlanLine(state));
  } catch {
    return undefined;
  }
}

/** "Local mode" / "user@example.com · 12m left" — real, from the session. */
export function sessionStatusLineFor(runtime: Runtime): string {
  const stored = loadSession();
  if (!stored || sessionExpired(stored)) return 'Local mode';
  return stored.user.email;
}

/** Seconds left in the current free session (display only; server enforces). */
export function sessionCountdown(): string | null {
  const stored = loadSession();
  if (!stored || sessionExpired(stored)) return null;
  const state = RemoteProvider.sharedSession();
  if (!state?.sessionExpiresAt) return null;
  return formatCountdown(sessionSecondsLeft(state.sessionExpiresAt));
}

File: grace/cli/tui/info.py
"""TUI runtime facts (port of src/cli/tui/info.ts).

Every value in the header/home screen comes from real state: the workspace
the CLI was launched in, the configured provider/model, the auth session and
(best-effort) the server's free-plan quota. Nothing is invented.
"""

import os

from grace.auth.client import ApiClient
from grace.auth.session import load_session, session_expired
from grace.cli.free_plan import (
    banner_free_plan_line,
    format_countdown,
    session_seconds_left,
)
from grace.cli.ui.theme import strip_ansi
from grace.meta import VERSION
from grace.providers.remote import RemoteProvider
from grace.util_text import short_path


def build_tui_info(runtime, free_plan_line: str | None = None) -> dict:
    """Synchronous snapshot of real runtime facts."""
    stored = load_session()
    logged_in = stored is not None and not session_expired(stored)
    session_label = stored["user"]["email"] if logged_in and stored else "Local mode"

    served = runtime.provider.server_provider if hasattr(runtime.provider, "server_provider") and isinstance(runtime.provider, RemoteProvider) else None
    provider_label = (served or {}).get("label") or (runtime.provider.label if runtime.provider else "")
    model = runtime.provider.get_model().id if runtime.provider else ""

    return {
        "version": VERSION,
        "workspace": short_path(runtime.root, os.path.expanduser("~")),
        "provider": provider_label,
        "providerAvailable": runtime.provider is not None,
        "providerError": runtime.provider_error or None,
        "model": model,
        "session": session_label,
        "freePlan": free_plan_line,
    }


def refresh_free_plan() -> str | None:
    """Best-effort free-plan + session-time fetch from the backend (real data).
    Never delays or breaks the UI: failures just leave the line unset."""
    stored = load_session()
    if not stored or session_expired(stored):
        return None
    try:
        state = ApiClient(stored["apiUrl"], 2000).get_usage(stored["token"])
        # Seed the shared session view so the live countdown renders even
        # before the first task (display only — the server enforces).
        RemoteProvider.set_shared_session(state)
        return strip_ansi(banner_free_plan_line(state))
    except Exception:
        return None


def session_status_line_for(runtime) -> str:
    stored = load_session()
    if not stored or session_expired(stored):
        return "Local mode"
    return stored["user"]["email"]


def session_countdown() -> str | None:
    stored = load_session()
    if not stored or session_expired(stored):
        return None
    state = RemoteProvider.shared_session()
    if not state or not state.get("sessionExpiresAt"):
        return None
    return format_countdown(session_seconds_left(state["sessionExpiresAt"]))
